### PR TITLE
[RFC] driver: add a voltage regulator framework

### PR DESCRIPTION
**Pull request by etienne-lms**
Proposal for a voltage regulator driver framework.

### CODE REVIEW

File: core/drivers/regulator/regulator_gpio.c [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+		return TEE_ERROR_GENERIC;
+	}
+
+	res = dt_driver_device_from_node_idx_prop("gpios", fdt, node, 0,
+						  DT_DRIVER_GPIO, gpio_ref);
+	if (res)
+		return res;
+
+	cuint = fdt_getprop(fdt, node, "states", &len);
+	if (!cuint || len != 4 * sizeof(fdt32_t)) {
```
Comment (GseoC) [inline core/drivers/regulator/regulator_gpio.c:184]: len != 2 * sizeof(uint32_t)?

File: core/drivers/regulator/regulator_dt.c [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+	if (!provider)
+		return TEE_ERROR_DEFER_DRIVER_INIT;
+
+	*regulator = dt_driver_provider_priv_data(provider);
+	assert(*regulator);
+
+	return TEE_SUCCESS;
+}
+
+/* Helper function to regulator a regulator provider instance */
```
Comment (etienne-lms) [inline core/drivers/regulator/regulator_dt.c:141]: "... to **~~regulator~~ register**  a regulator ..."

File: core/drivers/regulator/regulator_dt.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t phandle = 0;
+
+	phandle = fdt_get_phandle(fdt, node);
+	switch (phandle) {
+	case 0:
+		/* We can ignore regulators without any phandle */
+		return TEE_SUCCESS;
+	case (uint32_t)-1:
+		DMSG("Failed to find provide phandle");
```
Comment (etienne-lms) [inline core/drivers/regulator/regulator_dt.c:154]: s/provide/provider/

File: core/drivers/regulator/regulator_dt.c [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+		free(pending);
+
+		supplied = true;
+	}
+
+	return supplied;
+}
+
+/*
+ * Attempt to register pending regulators once their supply if found.
```
Comment (etienne-lms) [inline core/drivers/regulator/regulator_dt.c:235]: s/if/is/

File: core/drivers/regulator/regulator_gpio.c [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	int len = 0;
+
+	/*
+	 * DT bindings allows more than 1 GPIO to controller more than
+	 * 2 voltage levels. As it's not used so far in known platform
+	 * this implementation is simplified to support only 2 voltage
+	 * levels controlled with a single GPIO.
+	 */
+	if (dt_driver_device_from_node_idx_prop("gpios", fdt, node, 1,
+						DT_DRIVER_GPIO, gpio_ref) !=
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator_gpio.c:172]: This call is not expected to succeed. It is does, it means the regulator voltage level is controlled with at least 2 GPIOs whereas the implementation supports only 1 GPIO for voltage control (see commit message).
So I think the implementation is OK.

File: core/drivers/regulator/regulator_gpio.c (hunk 1 above)
Comment (etienne-lms) [reply core/drivers/regulator/regulator_gpio.c:184]: We expect to support only 2 voltage levels (controlled from a single GPIO pin). So property "states" should content 4 32bit cells: [VOLTAGE_LEVEL1 GPIO_LEVEL1 VOLTAGE_LEVEL2 GPIO_LEVEL2 ]. For example:
`states = <1800000 0x1>, <2900000 0x0>;`

File: core/drivers/regulator/regulator_gpio.c [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+	res = get_voltage_level_gpio(fdt, node, regu);
+	if (res)
+		goto err;
+
+	if (fdt_getprop(fdt, node, "vin-supply", NULL))
+		supply_name = "vin";
+
+	desc = (struct regu_dt_desc){
+		.name = regu_name,
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator_gpio.c:248]: Right, I missed that, I'll change `regulator_dt_register()` to `strdup()` the regulator name from `struct regu_dt_desc` only if needed. Thanks.

File: core/drivers/regulator/regulator_gpio.c [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	res = regulator_dt_register(fdt, node, node, &desc);
+	if (res) {
+		EMSG("Can't register regulator %s: %#"PRIx32, regu_name, res);
+		goto err;
+	}
+
+	return TEE_SUCCESS;
+
+err:
+	free(regu_name);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator_gpio.c:263]: Not needed, `free(NULL)` is a valid call.

File: core/drivers/regulator/regulator_dt.c [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+static TEE_Result parse_dt(const void *fdt, int node,
+			   struct regulator *regulator)
+{
+	struct regulator_property *fp = NULL;
+	const fdt32_t *cuint = NULL;
+	int len = 0;
+
+	FMSG("Regulator: parse DT node %s", fdt_get_name(fdt, node, NULL));
+
+	regulator->name = fdt_getprop(fdt, node, "regulator-name", NULL);
```
Comment (etienne-lms) [inline core/drivers/regulator/regulator_dt.c:280]: to fix, as reported by @GseoC. If "regulator-name" property name not found, preserved the name passed in `struct regu_dt_desc`.

**Comment by etienne-lms** [comment]
comments addressed.

**Comment by etienne-lms** [comment]
Maybe I should create smaller P-Rs to ease review of this series?

File: core/drivers/regulator/regulator.c [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	FMSG("%s %duV", regulator_name(regulator), level_uv);
+
+	if (level_uv < regulator->min_uv || level_uv > regulator->max_uv)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!regulator->ops->set_voltage) {
+		if (level_uv == regulator->cur_uv)
```
Comment (etienne-lms) [inline core/drivers/regulator/regulator.c:194]: could be test at line 192.

File: core/drivers/regulator/regulator.c [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+	FMSG("%s refcount: %u", regulator_name(regulator), regulator->refcount);
+
+	unlock_regulator(regulator);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result regulator_enable(struct regulator *regulator)
+{
+	assert(regulator);
```
Comment (GseoC) [inline core/drivers/regulator/regulator.c:110]: how about a non-assert check here? I'm not a real fan of it when it's not related to device-tree or boot-only checks.

Maybe you consider there's no need to do so, you tell me :)

Few occurrences in this file

**Comment by GseoC** [comment]
Can you squash review patches to ease review please ? :)

File: core/drivers/regulator/regulator.c (hunk 10 above)
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:110]: It's not expected to call the API with a NULL reference.

File: core/drivers/regulator/regulator.c [hunk 11]
```diff
@@ -209,6 +209,81 @@ TEE_Result regulator_set_voltage(struct regulator *regulator, int level_uv)
...
+			if (ref_min < regulator->min_uv) {
+				int diff = regulator->min_uv - ref_min;
+				int incr = DIV_ROUND_UP(diff, ref_step);
+
+				ref_min += ref_step * incr;
+			}
+
+			if (ref_max > regulator->max_uv) {
+				int diff = ref_max - regulator->max_uv;
+				int decr = diff / ref_step;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:248]: It must be div-round-down (done here). We can't go over `regulator->max_uv`.

File: core/drivers/regulator/regulator.c [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+	int min_uv = 0;
+	int max_uv = 0;
+	int uv = 0;
+
+	if (!regulator || !regulator->ops || !regulator->ops->get_state ||
+	    !regulator->ops->set_state)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	regulator_get_range(regulator, &min_uv, &max_uv);
+	if (min_uv > max_uv)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:302]: yes.

File: core/include/drivers/regulator.h [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+ * Copyright (c) 2023, STMicroelectronics
+ */
+#ifndef DRIVERS_REGULATOR_H
+#define DRIVERS_REGULATOR_H
+
+#include <assert.h>
+#include <bitstring.h>
+#include <kernel/mutex.h>
+#include <sys/queue.h>
+#include <tee_api_types.h>
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/drivers/regulator.h:12]: thanks

File: core/include/drivers/regulator.h [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+struct regulator {
+	/* Fields initialized by caller of regulator_register() */
+	const struct regulator_ops *ops;
+	struct regulator *supply;
+	void *priv;
+	const char *name;
+	int min_uv;
+	int max_uv;
+	/* Fields internal to regulator framework */
+	int cur_uv;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/drivers/regulator.h:76]: It could but I think it's fine and easier as-is.

**Review by etienne-lms — COMMENTED** [review]
I'll fix the issues and squash all fixup commits.

File: core/drivers/regulator/regulator.c (hunk 10 above)
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:110]: I have ensured `assert(regulator)` is called at entry of each regulator framework API functions.

File: core/drivers/regulator/regulator.c [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+		EMSG("Can't disable %s", regulator_name(regulator->supply));
+		panic();
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result regulator_disable(struct regulator *regulator)
+{
+	assert(regulator);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:157]: preserved

File: core/drivers/regulator/regulator.c (hunk 9 above)
Comment (etienne-lms) [reply core/drivers/regulator/regulator.c:194]: addressed

**Comment by etienne-lms** [comment]
P-R https://github.com/OP-TEE/optee_os/pull/6326 merged with bugs, fixed by https://github.com/OP-TEE/optee_os/pull/6328.

DT support for regulator configuration: https://github.com/OP-TEE/optee_os/pull/6329

**Comment by etienne-lms** [comment]
Almost all changes were merged:
https://github.com/OP-TEE/optee_os/pull/6326
https://github.com/OP-TEE/optee_os/pull/6328
https://github.com/OP-TEE/optee_os/pull/6329
https://github.com/OP-TEE/optee_os/pull/6339
https://github.com/OP-TEE/optee_os/pull/6342
https://github.com/OP-TEE/optee_os/pull/6343
https://github.com/OP-TEE/optee_os/pull/6344
https://github.com/OP-TEE/optee_os/pull/6345

Still on-going:
https://github.com/OP-TEE/optee_os/pull/6345 (regulator-gpio)
+ the Test PTA related to regulator and DT

**Pull request closed without merging**